plugins/chat/app/services/chat/create_category_channel.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for creating a new category chat channel.
+  #
+  # @example
+  #  Service::Chat::CreateCategoryChannel.call(
+  #   guardian: guardian,
+  #   name: "SuperChannel",
+  #   description: "This is the best channel",
+  #   slug: "super-channel",
+  #   category_id: category.id,
+  #  )
+  #
+  class CreateCategoryChannel
+    include Service::Base
+
+    # @!method call(guardian:, **params_to_create)
+    #   @param [Guardian] guardian
+    #   @param [Hash] params_to_create
+    #   @option params_to_create [String] name
+    #   @option params_to_create [String] description
+    #   @option params_to_create [String] slug
+    #   @option params_to_create [Boolean] auto_join_users
+    #   @option params_to_create [Integer] category_id
+    #   @return [Service::Base::Context]
+
+    policy :can_create_channel
+    contract
+    step :set_auto_join_users_default
+    policy :category_channel_does_not_exist
+    model :category, :fetch_category
+    model :channel, :create_channel
+    model :membership, :create_membership
+    step :enforce_automatic_channel_memberships
+
+    # @!visibility private
+    class Contract
+      attribute :name, :string
+      attribute :description, :string
+      attribute :slug, :string
+      attribute :category_id, :integer
+      attribute :auto_join_users, :boolean, default: false
+
+      validates :category_id, presence: true
+
+      validates :name, length: { maximum: SiteSetting.max_topic_title_length }
+    end
+
+    private
+
+    def set_auto_join_users_default(contract:, **)
+      contract.auto_join_users =
+        ActiveRecord::Type::Boolean.new.deserialize(context[:auto_join_users]) || false
+    end
+
+    def can_create_channel(guardian:, **)
+      guardian.can_create_chat_channel?
+    end
+
+    def fetch_category(contract:, **)
+      Category.find_by(id: contract.category_id)
+    end
+
+    def category_channel_does_not_exist(contract:, **)
+      !Chat::Channel.exists?(
+        chatable_type: "Category",
+        chatable_id: contract.category_id,
+        name: contract.name,
+      )
+    end
+
+    def create_channel(category:, contract:, **)
+      category.create_chat_channel!(
```

Comment:
As we’re calling this step using `model`, we should probably use the non-bang method to create it. (if we want to leverage the fact the service will fail if the model isn’t created)